### PR TITLE
Fix ps0 OOM when workers too many.

### LINKED ISSUE #22137: ps0 will be OOM using MonitoredTrainingSession when workers too many
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.4
- **Python version**: 2.7.5
- **Bazel version (if compiling from source)**: 1.9
- **GCC/Compiler version (if compiling from source)**: 4.9.2
- **CUDA/cuDNN version**: CPU
- **GPU model and memory**: CPU
- **Exact command to reproduce**: Using MonitoredTrainingSession and thousands of workers.

### Describe the problem
My model is large and complex. Thousands of worker&ps are needed and MonitoredTrainingSession is used. In the stage of initialization, the memory of ps0 will raise up to 100G and then OOM, but other ps work correctly. Op device placement is tf.train.replica_device_setter and no other special strategy.

I open the log_device_placement and found that all ops related with report_uninitialized_xxx are placed on ps0. I think this is the root cause.

I change the code as PR #22136 , and then it works correctly.

### PR DESCRIPTION
**Pull request by jackonan**
When training with MonitoredTrainingSession in distributed mode,
all ops related with report_uninitialized_xxx will be placed on ps0,
which may lead to OOM when workers up to thousands.

We use a environment variable to solve it. When OOM happens, users can
set the local device to fix the problem, such as
os.environ['TF_LOCAL_DEVICE'] = '/job:worker/task:index'.

Fixes #22137

### CODE REVIEW

File: tensorflow/python/ops/resources.py [hunk 1]
```diff
@@ -86,7 +87,8 @@ def report_uninitialized_resources(resource_list=None,
     resource_list = shared_resources() + local_resources()
   with ops.name_scope(name):
     # Run all operations on CPU
-    with ops.device("/cpu:0"):
+    local_device = os.environ.get("TF_LOCAL_DEVICE", "/cpu:0")
```
Comment (ispirmustafa) [inline tensorflow/python/ops/resources.py]: TF_LOCAL_DEVICE seems a generic name. It may collide with some existing users. 
Is it possible to find a more specific name, something like 'TF_DEVICE_REPORT_UNINITIALIZED'.

Could you please also give us info about whether you have run this and got problem solved?

Comment (jackonan) [reply tensorflow/python/ops/resources.py]: Here is the issue. https://github.com/tensorflow/tensorflow/issues/22137

Comment (ispirmustafa) [reply tensorflow/python/ops/resources.py]: So I assume this PR will fix that issue.

Comment (jackonan) [reply tensorflow/python/ops/resources.py]: Yes.  "with ops.device('/cpu:0')" also appears at some other places, such as 'tensorflow/python/estimator/estimator.py'. I think it will lead to OOM when the workers are too many, but I do not use it in my case.

Comment (jackonan) [reply tensorflow/python/ops/resources.py]: @ispirmustafa TF_DEVICE_REPORT_UNINITIALIZED seems too special for users and they may not know what it means.

Comment (martinwicke) [reply tensorflow/python/ops/resources.py]: TF_LOCAL_DEVICE is very generic, and I don't think users will know what it means either.

In particular, since this is not documented, nobody will know what this does without looking through the code. I'm not sure we necessarily want to widely broadcast this, but I agree with @ispirmustafa that a more descriptive name for the environment variable would be better. Maybe TF_DEVICE_FOR_UNINITIALIZED_VARIABLE_REPORTING? Since that's all it's used for.

Comment (jackonan) [reply tensorflow/python/ops/resources.py]: @martinwicke @ispirmustafa I agree with you in some way. The root cause of this bug is device setting. When only '/cpu:0' is assigned, the runtime will pick the first job name and task index in alphabet order to fill the device name. That is why '/job:ps/task:0/cpu:0' was used when running in ps-worker mode. I can not evaluate what is the impact when the default device placement is changed. So I just fix the problem case by case.

**Comment by jackonan** [comment]
@martinwicke

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by ispirmustafa — DISMISSED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
Something is wrong with the sync infra. This should have merged but didn't, I'm looking into it.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Weird, this PR should have been merged by that merge. Not sure what happened.
